index.js

```diff
@@ -2,58 +2,132 @@
+const { promisify } = require('util')
 const readPackage = require('read-package-json')
-const builtins = require('builtins')()
+const builtins = require('module').builtinModules
 const resolveModule = require('resolve')
 const debug = require('debug')('dependency-check')
 const isRelative = require('is-relative')
 const globby = require('globby')
 const micromatch = require('micromatch')
+const pkgUp = require('pkg-up')
 
-const promisedReadPackage = function (pkgPath) {
-  return new Promise((resolve, reject) => {
-    readPackage(pkgPath, (err, pkg) => {
-      if (err) return reject(err)
-      resolve(pkg)
-    })
+const promisedFsAccess = promisify(fs.access)
+const promisedReadPackage = promisify(readPackage)
+
+async function resolveGlobbedPath (entries, cwd) {
+  if (typeof entries === 'string') entries = [entries]
+
+  debug('globby resolving', entries)
+
+  const resolvedEntries = await globby(entries, {
+    cwd,
+    absolute: true,
+    expandDirectories: false
   })
+
+  const paths = Object.keys(resolvedEntries.reduce((result, entry) => {
+    // Globby yields unix-style paths.
+    const normalized = path.resolve(entry)
+
+    if (!result[normalized]) {
+      result[normalized] = true
+    }
+
+    return result
+  }, {}))
+
+  debug('globby resolved', paths)
+
+  return paths
 }
 
-module.exports = function (opts, cb) {
-  let pkgPath = opts.path
-  const result = promisedReadPackage(pkgPath)
-    .catch(err => {
-      if (err && err.code === 'EISDIR') {
-        pkgPath = path.join(pkgPath, 'package.json')
-        return promisedReadPackage(pkgPath)
-      }
-      return Promise.reject(err)
-    })
-    .then(pkg => parse({
-      path: pkgPath,
-      package: pkg,
-      entries: opts.entries,
-      noDefaultEntries: opts.noDefaultEntries,
-      builtins: opts.builtins,
-      extensions: getExtensions(opts.extensions, opts.detective)
-    }))
-
-  if (cb) {
-    result
-      .then(value => { cb(null, value) })
-      .catch(err => { cb(err) })
-    return
+async function resolveModuleTarget (targetPath) {
+  let pkgPath, pkg
+
+  try {
+    pkg = await promisedReadPackage(targetPath)
+    pkgPath = targetPath
+  } catch (err) {
+    if (targetPath.endsWith('/package.json') || targetPath === 'package.json') {
+      throw new Error('Failed to read package.json: ' + err.message)
+    }
+
+    if (err && err.code === 'EISDIR') {
+      // We were given a path to a module folder
+      pkgPath = path.join(targetPath, 'package.json')
+      pkg = await promisedReadPackage(pkgPath)
+    }
   }
 
-  return result
+  if (!pkg) return undefined
+
+  return {
+    pkgPath,
+    pkg
+  }
+}
+
+async function resolveEntryTarget (targetPath) {
+  // We've been given an entry path pattern as the target rather than a package.json or module folder
+  // We'll resolve those entries and then finds us the package.json from the location of those
+  const targetEntries = await resolveGlobbedPath(targetPath)
+
+  if (!targetEntries[0]) {
```

Comment:
Can we check `.length`? It's a little unintuitive to check the existence of the first value.

Reply:
Sure! This was a micro-optimization I guess + it's actually the first item it is going to look for the package.json from